### PR TITLE
[skip ci] #27405: Update TT-NN programming guide

### PR DESCRIPTION
**Pull request by omilyutin-tt**
### Ticket
[#27405](https://github.com/tenstorrent/tt-metal/issues/27405)

Changes:
1. Make it clear what level of abstraction `MeshDevice` provides.
2. Comparison with pytorch, using code snippets.
3. Use new visualization API and ND placement-based sharding
4. Rename T3000 -> QuietBox (Wormhole) and TG -> Galaxy.

### CODE REVIEW

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 1]
```diff
@@ -631,3 +665,108 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
+input_b = ttnn.from_torch(
+    torch_input_b,
+    layout=ttnn.TILE_LAYOUT,
+    device=mesh_device,
+    mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=3),
+)
+
+# Perform matrix multiplication - done in parallel on 2 devices
+output = ttnn.matmul(input_a, input_b)
+
```
Comment (omilyutin-tt) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:754]: TODO: need to set up a call to all gather

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 2]
```diff
@@ -631,3 +665,111 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
 APIs will be further refined in future releases. A proposal for refined set of APIs can be found [here](https://github.com/tenstorrent/tt-metal/issues/13852).
+
+
+## 9. TT-NN MeshDevice vs PyTorch Multi-Device Programming
+
+### 9.1 Overview of Multi-Device Handling
+
+This section compares TT-NN's MeshDevice with PyTorch's approach to **single-node multi-device** programming. Note that PyTorch Distributed (torch.distributed) is designed for multi-node systems and is not the appropriate comparison here. Similarly, TT-NN's multi-node MeshDevice support is currently under development and not covered in this comparison.
+
+**MeshDevice in TT-NN (Single-Node):**
```
Comment (omilyutin-tt) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:690]: Note I am deliberately avoiding multi host distributed for now.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 3]
```diff
@@ -631,3 +665,111 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
+- **Manual Management**: Explicitly place tensors on specific devices (`cuda:0`, `cuda:1`, etc.) and manually orchestrate data movement and synchronization between devices
+- **DataParallel**: Use `torch.nn.DataParallel` which automatically splits data across GPUs but operates in a single process with potential GIL bottlenecks
+- **DistributedDataParallel**: While primarily designed for multi-node, DDP can be used on single-node with multiple processes (one per GPU) for better performance than DataParallel
+
+For manual multi-GPU management in a single process, PyTorch provides no built-in abstractions for:
+- Coordinated command dispatch across devices
+- Efficient collective communication operations
+- Automatic kernel or data broadcasting
+
+### 9.2 Code Comparison: Matrix Multiplication with All-Gather
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:710]: could we do this as a side-by-side compare in markdown? https://gist.github.com/nottheswimmer/7837eec10fac0d1197fc5a8bfc0b96f3

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 4]
```diff
@@ -631,3 +665,111 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
+```python
+import ttnn
+import torch
+
+# Open a 1x2 MeshDevice
+mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 2))
+
+# Create input tensors
+torch_input_a = torch.randn(1, 1, 128, 128, dtype=torch.bfloat16)
+torch_input_b = torch.randn(1, 1, 128, 256, dtype=torch.bfloat16)
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:734]: does `ttnn.rand` work now?

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 5]
```diff
@@ -631,3 +665,111 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
+
+# Replicate input A and shard input B (each device gets 128x128)
+input_a = ttnn.from_torch(
+    torch_input_a,
+    layout=ttnn.TILE_LAYOUT,
+    device=mesh_device,
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:741]: Maybe make a comment to say that the `.to_device` is implicit in this API. I had someone ask me about this at some point as well.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 6]
```diff
@@ -631,3 +665,111 @@ See `models/demos/t3000/llama2_70b/tests/test_llama_perf_decode.py::test_Llama_p
...
+    mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=3),
+)
+
+# Perform matrix multiplication - done in parallel on 2 devices
+output = ttnn.matmul(input_a, input_b)
+
+# All-gather to collect results from all devices
+gathered = ttnn.all_gather(output, dim=3, cluster_axis=0)
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:757]: Maybe in the intro to this section, adding a descriptor about the tensor handle you get back and how that differs against individual single-device handles.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 7]
```diff
@@ -50,28 +53,41 @@ These concepts are key to understanding how we scale models using **Data-Paralle
 
 ## 2. MeshDevice
 
+MeshDevice is a virtual device abstraction that bundles together multiple physical devices to enable efficient parallel execution across a mesh topology. This abstraction is natively supported at the runtime level, allowing for deep integration with the hardware and dispatch mechanisms. When operations are dispatched to a MeshDevice, command queues are utilized to distribute work across all constituent devices in parallel, significantly reducing dispatch overhead compared to sequential device-by-device execution.
+
+To optimize performance, MeshDevice implements several key optimizations:
+- **Kernel Compilation Broadcasting**: When kernels are compiled for execution on a MeshDevice, the compilation artifacts are automatically broadcasted to all devices in the mesh where applicable, avoiding redundant per-device compilation.
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:59]: Maybe need to caveat when we go multi-node

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 8]
```diff
@@ -50,28 +53,41 @@ These concepts are key to understanding how we scale models using **Data-Paralle
...
+- **Data Broadcasting**: For replicated tensors (where the same data needs to exist on multiple devices), MeshDevice leverages the mesh topology to efficiently broadcast data across devices rather than performing individual writes to each device.
+- **Unified Command Dispatch**: Operations are dispatched through mesh-aware command queues that coordinate execution across all devices, ensuring synchronized parallel execution.
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md]: "synchronized" -> "lock-step"?

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md [hunk 9]
```diff
@@ -50,28 +53,41 @@ These concepts are key to understanding how we scale models using **Data-Paralle
...
+
+While MeshDevice provides these performance optimizations, it maintains explicit control over data distribution and communication patterns. MeshDevice does not hide the distributed nature of the computation - users must explicitly specify:
+- **Data Distribution**: How input tensors are distributed across devices (sharded or replicated) using mesh mappers
```
Comment (cfjchu) [inline tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:64]: Do you intend to add section on new placement APIs you added as well?

**Review by cfjchu — APPROVED** [review]
Looks great! 😍

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 5 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:741]: Done

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 3 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:710]: Yep, looks more clear

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 8 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md]: Done

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 9 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:64]: I sort of do here, not sure if it is worth a full doc on its own - maybe.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 7 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:59]: I think it is probably fine not to mention here, for simplicity. Will definitely elaborate on things like that in the dedicated section.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 4 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:734]: It does, but here I want to demonstrate input sharding / replication... Not sure what would be the best way to do that. I added a comment that rand can also be used here for random inputs.

File: tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md (hunk 6 above)
Comment (omilyutin-tt) [reply tech_reports/Programming_Mesh_of_Devices/Programming_Mesh_of_Devices_with_TT-NN.md:757]: Done, specified in a comment here
